go_stash/groups.go

```diff
@@ -0,0 +1,188 @@
+/*
+Copyright 2021 The Flux authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package gostash
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+)
+
+const (
+	groupsURI       = "admin/groups"
+	groupMembersURI = "admin/groups/more-members"
+)
+
+// GroupsService is a client for communicating with stash groups endpoint
+// Stash API docs: https://docs.atlassian.com/DAC/rest/stash/3.11.3/stash-rest.html
+type GroupsService service
+
+// Group represents a stash group.
+type Group struct {
+	// Session is the session object for the group.
+	Session Session `json:"sessionInfo,omitempty"`
+	// Name is the name of the group.
+	Name string `json:"name,omitempty"`
+	// Delete is the delete flag for the group.
+	Deleteable bool `json:"deletable,omitempty"`
+}
+
+// GroupList represents  a list of stash groups.
+type GroupList struct {
+	// Paging is the paging information.
+	Paging
+	// Groups is the list of stash groups.
+	Groups []*Group `json:"values,omitempty"`
+}
+
+// GetGroups returns a slice of groups.
+func (g *GroupList) GetGroups() []*Group {
+	return g.Groups
+}
+
+// List retrieves a list of stash groups.
+// Paging is optional and is enabled by providing a PagingOptions struct.
+// A pointer to a paging struct is returned to retrieve the next page of results.
+// List uses the endpoint "GET /rest/api/1.0/admin/groups".
+// The authenticated user must have the LICENSED_USER permission to call this resource.
+func (s *GroupsService) List(ctx context.Context, opts *PagingOptions) (*GroupList, error) {
+	query := addPaging(&url.Values{}, opts)
+	req, err := s.Client.NewRequest(ctx, http.MethodGet, newURI(groupsURI), *query, nil, nil)
+	if err != nil {
+		return nil, fmt.Errorf("list groups failed, %w", err)
```

Comment:
Generally you use `"message: %w"` when wrapping errors for example https://cs.opensource.google/go/go/+/refs/tags/go1.17.1:src/net/http/server.go;l=324